proposals/0000-hasfield-redesign.rst

```diff
@@ -0,0 +1,1266 @@
+the latter should first be added to the forthcoming ``ghc-experimental``
+package.  Thus it adds two new modules, ``GHC.Records`` and
+``GHC.Records.Experimental``.
+
+The ``GHC.Records`` module (in the ``base`` package) will be defined as follows::
+
+  {-# LANGUAGE AllowAmbiguousTypes #-}     -- for type of getField
+  {-# LANGUAGE FunctionalDependencies #-}  -- for HasField class
+
+  module GHC.Records
+    ( HasField(getField)
+    ) where
+
+  import GHC.Types (Constraint, TYPE)
+
+  -- | Constraint representing the fact that a field @x@ of type @a@ can be
+  -- selected from the record type @r@.
+  --
+  -- This will be solved automatically for built-in records where the field is
+  -- in scope, but manual instances may be provided as well.
+  --
+  type HasField :: forall {k} {r_rep} {a_rep} . k -> TYPE r_rep -> TYPE a_rep -> Constraint
+  class HasField x r a | x r -> a where
+    -- | Selector function to extract the field @x@ from the record @r@.
+    getField :: r -> a
+
+
+The ``GHC.Records.Experimental`` module (in the ``ghc-experimental`` package)
+will be defined as follows::
+
+  {-# LANGUAGE AllowAmbiguousTypes #-}     -- for type of setField
+  {-# LANGUAGE DefaultSignatures #-}       -- for setField/modifyField
+  {-# LANGUAGE FunctionalDependencies #-}  -- for SetField class
+
+  module GHC.Records.Experimental
+    ( HasField(getField)
+    , SetField(setField, modifyField)
+    , Field
+    ) where
+
+  import GHC.Records (HasField(getField))
+  import GHC.Types (Constraint, TYPE)
+
+  -- | Constraint representing the fact that a field @x@ of type @a@ can be
+  -- updated in the record type @r@.
+  --
+  -- This will be solved automatically for built-in records where the field is
+  -- in scope, but manual instances may be provided as well.
+  --
+  -- Instances of this class are subject to the following laws, for every record
+  -- value @r@ and field @x@:
+  --
+  -- > modifyField @x id r === r or ⊥
+  -- > (modifyField @x g . modifyField @x f) r === modifyField @x (g . f) r
+  -- > setField @x v r == modifyField @x (\ _ -> v) r
+  --
+  -- Where a 'HasField' instance is available as well as an instance of this
+  -- class, they must together satisfy the laws defined on 'Field'.
+  --
+  type SetField :: forall {k} {r_rep} {a_rep} . k -> TYPE r_rep -> TYPE a_rep -> Constraint
+  class SetField x r a | x r -> a where
+    -- | Change the value stored in the field @x@ of the record @r@.
+    modifyField :: (a -> a) -> r -> r
+    default modifyField :: (r_rep ~ LiftedRep, a_rep ~ LiftedRep, HasField x r a) => (a -> a) -> r -> r
+    modifyField f r = setField @x (f (getField @x r)) r
+
+    -- | Update function to set the field @x@ in the record @r@.
+    setField :: a -> r -> r
+    default setField :: a_rep ~ LiftedRep => a -> r -> r
+    setField v = modifyField @x (\ _ -> v)
+
+    {-# MINIMAL modifyField | setField #-}
```

Comment:
Is this really true? `modifyField` and `setField` don't always have default definitions.

Reply:
I don't think that's how `DefaultSignatures` works, but I haven't actually verified it experimentally...

Reply:
You indeed get a type error if the default signature doesn't match, so there is no problem here.